core/modules/loader/Central.h

```diff
@@ -0,0 +1,133 @@
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ *
+ */
+#ifndef LSST_QSERV_LOADER_CENTRAL_H_
+#define LSST_QSERV_LOADER_CENTRAL_H_
+
+// system headers
+#include <boost/bind.hpp>
+#include <boost/asio.hpp>
+#include <thread>
+#include <vector>
+
+// Qserv headers
+#include "loader/ClientServer.h"
+#include "loader/MasterServer.h"
+#include "loader/MWorkerList.h"
+#include "loader/WWorkerList.h"
+#include "loader/WorkerServer.h"
+#include "proto/loader.pb.h"
+#include "util/ThreadPool.h"
+
+
+
+namespace lsst {
+namespace qserv {
+namespace loader {
+
+/// TODO add fileId and row to this so it can be checked in _workerKeyInsertReq
+struct ChunkSubchunk {
+    ChunkSubchunk(int chunk_, int subchunk_) : chunk(chunk_), subchunk(subchunk_) {}
+    int const chunk;
+    int const subchunk;
+    friend std::ostream& operator<<(std::ostream& os, ChunkSubchunk csc);
+};
+
+
+/// This class is 'central' to the execution of the program, and must be around
+/// until the bitter end. As such, it can be accessed by normal pointers.
+/// This class is central to loader workers and the master.
+/// It is the base class for CentralWorker, CentralMaster, and CentralClient
+class Central {
+public:
+    Central(boost::asio::io_service& ioService,
+            std::string const& masterHostName, int masterPort)
+        : _ioService(ioService), _masterAddr(masterHostName, masterPort),
+          _checkDoListThread([this](){ _checkDoList(); }){}
+
+    Central() = delete;
+
+    virtual ~Central();
+
+    void run();
+
+    std::string getMasterHostName() const { return _masterAddr.ip; }
+    int getMasterPort() const { return _masterAddr.port; }
+    NetworkAddress getMasterAddr() { return _masterAddr; }
+
+    uint64_t getNextMsgId() { return _sequence++; }
+
+    int getErrCount() const { return _server->getErrCount(); }
+
+    /// Send the contents of 'sendBuf' to 'host:port'. This waits for the message to be
+    /// sent before returning.
+    void sendBufferTo(std::string const& host, int port, BufferUdp& sendBuf) {
+        _server->sendBufferTo(host, port, sendBuf);
+    }
+
+    // Only allow tracked commands on the queue
+    void queueCmd(util::CommandTracked::Ptr const& cmd) {
+        _queue->queCmd(cmd);
+    }
+
+    bool addDoListItem(DoListItem::Ptr const& item) {
+        return _doList.addItem(item);
+    }
+
+    bool runAndAddDoListItem(DoListItem::Ptr const& item) {
+        _doList.runItemNow(item);
+        return _doList.addItem(item);
+    }
+
+
+    virtual std::string getOurLogId() { return "baseclass"; }
+
+protected:
+    /// Repeatedly check the items on the _doList.
+    void _checkDoList();
+
+    boost::asio::io_service& _ioService;
+
+    /// Initialization order is important.
+    DoList _doList{*this}; ///< List of items to be checked at regular intervals.
```

Comment:
passing `this` to another object during initialization is dangerous.

Reply:
What is there is safe, since the this pointer passed to doList wont be used until the last member '_checkDoListThread' is defined. This is somewhat fragile in that moving the  _checkDoListThread declaration can cause a race condition.

The alternative is to break construction into two parts. The regular constructor is called and then a second step, I'll call go(), is called, which sets this pointer and creates the thread and moves in into _checkDoListThread. I'm ok with either. Factory functions don't really work well with base classes, since the child constructors wont call them.